Enhance error handling for notifications from private blogs

justtwago opened the pull request:
Fixes #17127 

This PR handles errors for two types of notifications related to private blogs to which the user does not have access: comments and mentions. Mentions are split into mentions either in a comment or a post.

Moreover, the Reader's comments screen displayed an infinite loading progress bar, so it's been fixed too.

**To test:**

First, please temporarily leave a P2 from which you have tons of notifications to make it private for you. I assume it's cdRpT-p2.
Go to Notifications screen and test the following cases
- Open a "mention in/reply to a comment" notification that you've already seen (it has to be cached). You should see the comment on a Mention notification details screen
- Go to the comments screen from the details (tap on the "Comment on >" section on the top of the screen)
- [ ] Make sure you see an error message stating that you cannot access the private blog instead of an infinite loading state.
- [ ] Go back to the notifications screen and open a "mention in/reply to a comment" notification that is not cached yet. Make sure you can see a similar error without seeing the comment.
- Go back to the notifications and find a notification mentioning you **in a post**.
- [ ] Tap on it and ensure you see an error message stating you cannot access the private blog.
- [ ] Smoke test other notifications
- [ ] Subscribe back to the P2 and test the cases again to ensure you have access to all the notifications' details

## Regression Notes
1. Potential unintended areas of impact
Reader comments, Notification details screens

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests

3. What automated tests I added (or what prevented me from doing so)
There were no initial tests to the legacy code this PR touches

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

justtwago committed:
Display more informative errors for notification comment details screen

justtwago committed:
Add more informative error messages to notifications reader details

justtwago committed:
Fix infinite loading on reader comments screen

Display an informative error message instead

hassaanelgarem commented:
> User cannot access this private blog

Is this string returned from the endpoint, or is it defined locally?

alpavanoglu commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift:
```diff
@@ -284,18 +310,18 @@ private extension NotificationCommentDetailViewController {
                 completion(comment)
             }, failure: { error in
-                completion(nil)
+                failure(error)
```
Does this display BE error directly? If so, that won't be localised and may not make sense for the user.

salimbraksa commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationCommentDetailViewController.swift:
```diff
@@ -229,16 +232,39 @@ private extension NotificationCommentDetailViewController {
+                guard let self = self else {
+                    return
+                }
+                self.showErrorView(error: error)
```
[Nit] In the nested closures, we currently redeclare `[weak self]` after it has already been captured as `weak` by an outer closure. For conciseness and based on [this Stackoverflow question](https://stackoverflow.com/questions/38739129/do-capture-lists-of-inner-closures-need-to-redeclare-self-as-weak-or-unowne), we can safely omit the redundant `weak` capture in the inner closures while still avoiding strong reference cycles.

It's okey to keep the redundant `[weak self]`, but initially, I found the method hard to read. Upon a closer look, I realized that the repeated `[weak self]` captures contributed to its perceived complexity.

justtwago committed:
Fix ReaderDetailCoordinator tests

justtwago committed:
Use localized error message fetch likes from private blog error

justtwago committed:
Use localized error message fetch comment details from private blog error

justtwago committed:
Use localized error message fetch reader comments from private blog error

justtwago committed:
Use localized error message fetch reader detail from private blog error

justtwago commented:
Thank you folks! I fixed error messages to display localized strings except for the raw API error message. Also removed redundant weak self checks.

salimbraksa commented on WordPress/Classes/ViewRelated/Reader/Detail/ReaderDetailCoordinator.swift:
```diff
@@ -778,3 +790,17 @@ extension ReaderDetailCoordinator {
+    struct Strings {
+        static let fetchDetailFromPrivateBlogErrorMessage = NSLocalizedString(
+            "readerDetailCoordinator.readerDetail.privateBlogErrorMessage",
+            value: "You have no access to the private blog.",
```
[Nit] How about rewording the error message to "You don't have permission to view this private blog"?

![](https://github.com/wordpress-mobile/WordPress-iOS/assets/9609223/5e804bf8-12c9-41b0-933e-1b85391570fa)

salimbraksa commented on WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.m:
```diff
@@ -791,15 +792,26 @@ - (void)refreshNoResultsView
+        if (error && [error.domain isEqualToString:WordPressComRestApiErrorDomain] && error.code == WordPressComRestApiErrorCodeAuthorizationRequired) {
+            subtitle = NSLocalizedString(@"You have no access to the private blog.",
+                                          @"Error message that informs reader comments from a private blog cannot be fetched.");
+
```
[Suggestion] I think this localized string should use the reverse dns format, right?

justtwago replied:
I'm unsure if I can put a key, value, and description to the Objective-C code. All the `NSLocalizedString` usages I found in our app written in Objective-C look similar to this one.

salimbraksa reviewed (approved):
I've dropped some minor comments, but PR works as expected. 👍

justtwago committed:
Modify the blog access error message copy

You have no access to the private blog. -> You don't have permission to view this private blog.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/274">24.5</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
